pcdsdevices/interface.py

```diff
@@ -217,9 +218,162 @@ def __init_subclass__(cls, **kwargs):
+            TimeoutError if anything does not connect during the timeout
+            duration.
+            Defaults to False.
+        timeout : float, optional
+            The maximum time to wait for connections if wait_connected
+            is True. Defaults to 0.005s per component, or at least 1s.
+        """
+        if cpts is None:
+            activation_call = True
+            try:
+                cpts: Iterable[Component] = self._sig_attrs.values()
+            except AttributeError:
+                # No _sig_attrs means not a Device, meaning no cpts
+                self._active_flag = True
+                return
+            if wait_connected and timeout is None:
+                timeout = max(1.0, 0.005 * len(list(self.walk_components())))
+        else:
+            activation_call = False
+        if wait_connected and timeout is None:
+            timeout = max(1.0, 0.005 * len(cpts))
+        start = time.monotonic()
+        objs: List[Union[Device, Signal]] = []
+        for cpt in cpts:
+            if isinstance(cpt, str):
+                cpt: Component = self._sig_attrs[cpt]
+            if any((
+                cpt._subscriptions and ensure_subs,
+                include_lazy,
+                not cpt.lazy,
+            )):
+                objs.append(getattr(self, cpt.attr))
+        if activation_call:
+            self._active_flag = True
+        for obj in objs:
+            try:
+                obj.ensure_cpts()
+            except AttributeError:
+                pass
+            if not obj.connected and wait_connected:
+                elapsed = time.monotonic() - start
+                remaining = timeout - elapsed
+                if remaining <= 0:
+                    raise TimeoutError(
+                        'Timeout waiting for cpts to connect.'
+                    )
+                obj.wait_for_connection(timeout=remaining)
+
+    def destroy(self):
+        """
+        Disconnect and destroy all loaded signals and subdevices.
+
+        This is modified from Device.destroy() to consider that some
+        non-lazy signals may not have been created yet, else we risk
+        creating them during the destroy process.
+        """
+        self._destroyed = True
+        exceptions = []
+        for attr, obj in self._signals.items():
+            try:
+                obj.destroy()
+            except Exception as ex:
+                ex.obj = obj
+                ex.attr = attr
+                exceptions.append(ex)
+
+        if exceptions:
+            if len(exceptions) == 1:
+                raise exceptions[0]
+            msg = ", ".join(
+                "{} ({})".format(ex.attr, ex.__class__.__name__)
+                for ex in exceptions
+            )
+            raise ExceptionBundle(
+                "Failed to disconnect all signals ({})".format(msg),
+                exceptions=exceptions,
+            )
+        self.unsubscribe_all()
```

Comment:
I know this is how it is in ophyd.Device (`super().destroy()` -> `OphydObject.destroy()` -> `self.unsubscribe_all()`), but I found it interesting that if an exception is raised the unsubscribe_all call would get missed

Reply:
You're 100% right. Let's optimize a bit further because we can and clean up as much as possible.